gapic-generator/lib/gapic/path_pattern/parser.rb

```diff
@@ -15,67 +15,102 @@
+      # @return [Gapic::PathPattern::Pattern]
+      def self.parse path_pattern
+        starts_with_slash = path_pattern.start_with? "/"
+        ends_with_slash = path_pattern.end_with? "/"
 
-      protected
+        remainder = path_pattern
+        remainder[0] = "" if starts_with_slash
+        remainder[-1] = "" if ends_with_slash
 
-      def parse! path_pattern
-        # segments contain either Strings or segment objects
         segments = []
-        segment_pos = 0
+        position = 0
+        until remainder.empty?
+          segment, position, remainder = parse_first_segment_with_position remainder, position
+          segments << segment
+        end
 
-        while (match = PATH_PATTERN.match path_pattern)
-          # The String before the match needs to be added to the segments
-          segments << match.pre_match unless match.pre_match.empty?
+        Pattern.new path_pattern, segments
+      end
 
-          segment, segment_pos = segment_and_pos_from_match match, segment_pos
-          segments << segment
+      # @private
+      # rubocop:disable Metrics/AbcSize
+      # rubocop:disable Metrics/MethodLength
+      # rubocop:disable Style/RegexpLiteral
+      def self.parse_first_segment_with_position url_pattern, position
+        # check for the w ildcard capture -- either * or **
+        wildcard_capture_regex = /^(?<pattern>\*\*|\*)(?:\/|$)/
+        if wildcard_capture_regex.match? url_pattern
+          match = wildcard_capture_regex.match url_pattern
+
+          wildcard_pattern = match[:pattern]
 
-          path_pattern = match.post_match
+          segment = PositionalSegment.new position, wildcard_pattern
+          remainder = match.post_match
+          return segment, position + 1, remainder
         end
 
-        # Whatever String is unmatched needs to be added to the segments
-        segments << path_pattern unless path_pattern.empty?
+        # check for the complex resource id segment --
+        # {<name_first>}<separator>{<name_second>} etc, e.g. {foo}-{bar}_{baz}
+        # see AIP-4231 Parsing resource names, Complex resource ID path segments
+        complex_resource_id_regex =
+          /^(?<segment_pattern>{(?<name_first>[^\/}]+?)}(?:(?<separator>[_\-~\.]){(?<name_seq>[^\/}]+?)})+)(?:\/|$)/
+        if complex_resource_id_regex.match? url_pattern
+          match = complex_resource_id_regex.match url_pattern
+          segment_pattern = match[:segment_pattern]
 
-        segments
-      end
+          resource_name_regex = /{(?<name>[^\/}]+?)}/
+          resource_names = segment_pattern.scan(resource_name_regex).flatten
+
+          segment = ResourceIdSegment.new :complex_resource_id, segment_pattern, resource_names
+          remainder = match.post_match
+          return segment, position, remainder
+        end
+
+        # check for the simple resource id segment -- {<name>},
+        # e.g. {foo} with an optional pattern, e.g. {foo=**} or {foo=bar}
```

Comment:
Are complex resource IDs and resource patterns mutually exclusive? I would expect not (although the AIP doesn't seem to be clear on this).

Reply:
The AIPs do not mention resource id patterns at all I think. My judgment here was that with complex ids being legacy feature and not recommended for the future use we are unlikely to see the patterns combined with them in practice. Therefore I valued not complicating the regex further.